### PR TITLE
new feature: slideshow duplication

### PR DESCRIPTION
**Pull request by ghost**
This new feature aims to implement the slideshow duplication: slideshow + relate images.
I took Tony Botalla's code (form duplication) as an example, so I finally got almost the same code. 

To avoid code duplication we could maybe try to make abstract it, modifying noviusos_form and \Nos\Controller_Admin_Crud, maybe more, I am not sure. Maybe It does not worth it...

Any thoughts will be appreciated.

#25115

### CODE REVIEW

**Comment by Meroje** [comment]
I don't know what to do of the `Model_Image::slidimg_link_to_page_id` property, maybe it could be updated if there is a translation of the page in the new context ? Maybe it should be up to the user too, just wanted to bring it up.

**Comment by ghost** [comment]
@Meroje  yes, I think it could make sense. Right now we keep the existent **page_id**, which it does not match any page belonging to the new context.

So we can proceed like this:

When saving the slideshow, if any of the given images has **slidimg_link_to_page_id**, we request **nos_page.page_id** by **page_context_common_id** and **page_context**. If we find it, it means that there is a translated page, otherwise we use null because, unless I am mistaken, it does not make sense to link a page of another context. 

Let me know what do you think.

**Comment by Meroje** [comment]
Looks fine to me but I would wait on input from others on the team

**Comment by shaoshiva** [comment]
You can easily find an item in another context like this :
```php
// Gets the linked page
$page = Model_Page::find($image->slidimg_link_to_page_id);
if (!empty($page)) {
    // Gets the linked page in the duplicate context
    $page = $page->find_context($duplicateContext) ?: $page;
}
```

**Comment by ghost** [comment]
@Meroje @shaoshiva You'll find a new version including the update of the linked page id. Let me know what do you think.

**Review by Meroje — APPROVED** [review]

File: classes/model/slideshow.model.php [hunk 1]
```diff
@@ -134,6 +135,27 @@ protected function duplicateSlideImages(Model_Slideshow $slider, Model_Slideshow
              */
             $clonedImage = clone $image;
             $clonedImage->slidimg_slideshow_id = $duplicatedSlider->slideshow_id;
+
+            /*
+             * For any given image containing a link to another site's page
+             * we find out if a translated page already exists on the target context.
+             * If so, we update the id of the linked page before save.
+             */
+            if ($clonedImage->slidimg_link_to_page_id) {
```
Comment (shaoshiva) [inline classes/model/slideshow.model.php]: I just noticed that there is a `page` relation on `Model_Image`. So you can do it in 3 lines :
```php
if ($clonedImage->page) {
    $clonedImage->page = $clonedImage->page->find_context($duplicatedSlider->slideshow_context) ?: $clonedImage->page;
}
```

Code not tested, but it should work.

**Review by shaoshiva — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
@shaoshiva 
It works fine, just two remarks:
1. I keep the assignment of NULL (it seems logic to me if there is no page in the new context)
2. I am not saving the new page in $clonedImage->page because It generates an error when saving the image "Primary key on model Nos\Page\Model_Page cannot be changed." I think that it tries to update the former page.

**Review by jguyomard — APPROVED** [review]

**Review by shaoshiva — APPROVED** [review]

**Pull request merged**